### ISSUE TITLE
Disable MAD mods

### ISSUE THREAD

**Issue by Alystrasz** [R2Northstar/NorthstarLauncher#670]
As we don't want MAD mods to conflict with other mods, mods that have been downloaded to join a server should be disabled after leaving said server.

**Comment by ASpoonPlaysGames** [comment]
In a similar vein as this: https://github.com/R2Northstar/NorthstarMods/issues/263

Any mods that get enabled/disabled specifically to join a server should get reverted back to the state they were in after leaving the server.

Some pointers for implementation:
- Can't just store this in UI vm, since `uiscript_reset` might get ran when enabling/disabling mods, so the data would be lost.
- Launcher implementation is probably best, perhaps saving enabledmods state via a script call and then rolling back to the saved state later via a second script call

**Comment by RoyalBlue1** [comment]
For implementation can't we enable/disable the mods according to the server without writing it to enabledmods and then on every server leave we load enabledmods again to reset to the previous state

**Comment by ASpoonPlaysGames** [comment]
> For implementation can't we enable/disable the mods according to the server without writing it to enabledmods and then on every server leave we load enabledmods again to reset to the previous state

Yeah that would work, could have "temporarily" enabled/disabled mods by doing that. However reloading mods on every server leave isn't great, we should check to make sure that there are currently some "temporarily" enabled/disabled mods first

**Comment by Alystrasz** [comment]
> For implementation can't we enable/disable the mods according to the server without writing it to enabledmods and then on every server leave we load enabledmods again to reset to the previous state

Can we really avoid using enabledmods here?
Maybe a server leave callback wouldn't be triggered if players rage quit the game during a match :smile:

**Comment by Alystrasz** [comment]
Launcher-side, is there any particular method (hook?) that is triggered on leaving a game?

**Comment by RegicidalBomb** [comment]
Is this what you're looking for? I think this might be it. So this is my first time collaborating on Northstar, so I'm not entirely sure how this works, but do you need a C++ function or a Squirrel function? I found this in the Respawn API (Squirrel).
```c++
void AddCallback_OnClientDisconnected( void functionref( entity player ) callbackFunc )
```
I'm assuming C++, I just don't know where to look, really. Will continue to search.

**Comment by Alystrasz** [comment]
> Is this what you're looking for? I think this might be it. So this is my first time collaborating on Northstar, so I'm not entirely sure how this works, but do you need a C++ function or a Squirrel function? I found this callback in the Respawn API (Squirrel).
> 
> ```c++
> void AddCallback_OnClientDisconnected( void functionref( entity player ) callbackFunc )
> ```

Thanks for taking a look!
Looks like `AddCallback_OnClientDisconnected` is a server-side function, and we're looking for a client-side function instead :/

**Comment by F1F7Y** [comment]
> Can we really avoid using enabledmods here? 

Yes

> Maybe a server leave callback wouldn't be triggered if players rage quit the game during a match 😄

Simply dont do anything on leave (apart from reloading mods) and only care about joining. If you dont touch enabledmods youll be a-okay.

**Comment by Alystrasz** [comment]
> > Maybe a server leave callback wouldn't be triggered if players rage quit the game during a match 😄
> 
> Simply dont do anything on leave (apart from reloading mods) and only care about joining. If you dont touch enabledmods youll be a-okay.

That's what I ended up doing in #758.

**Comment by Sandwhiched** [comment]
> Thanks for taking a look! Looks like `AddCallback_OnClientDisconnected` is a server-side function, and we're looking for a client-side function instead :/

Sorry, should have checked. Sorry for replying so late as well. I'm retiring and likely not coming back (emphasis on likely) because the vanilla (the only servers I played on, used Vanilla+) community have constantly been resolve-breakingly rude sometimes and I'm just drained of any motivation to continue playing or even develop mods/here. I could play on Northstar servers but again I'm just so drained. I might come back but it'll probably be a while.

**Comment by Sandwhiched** [comment]
Nah, I can't leave. I love this game too much. I was just really already mad yesterday and one guy in a match was just my final straw. This game's too good to abandon, and I was acting REALLY dramatic. Sorry :( I can look and try finding a server-side function, if that would help.

**Comment by Alystrasz** [comment]
> I can look and try finding a server-side function, if that would help.

Thanks for looking, but this issue should be solved once #758 is merged.

**Comment by Sandwhiched** [comment]
Got it! If you need any help let me know :D
